app/models/access.rb

```diff
@@ -0,0 +1,86 @@
+class Access < ApplicationRecord
+  ALLOWED_RESOURCES = %w[Organization FacilityGroup Facility].freeze
+  ACTION_TO_ROLE = {
+    view: [:manager, :viewer],
+    manage: [:manager]
+  }
+
+  belongs_to :user
+  belongs_to :resource, polymorphic: true, optional: true
+
+  enum role: {
+    viewer: "viewer",
+    manager: "manager",
+    super_admin: "super_admin"
+  }
+
+  validates :role, presence: true
+  validates :user, uniqueness: {scope: [:resource_id, :resource_type], message: "can only have one access per resource."}
+  validates :resource, presence: {unless: :super_admin?, message: "is required if not a super_admin."}
+  validates :resource, absence: {if: :super_admin?, message: "must be nil if super_admin"}
+  validates :resource_type, inclusion: {in: ALLOWED_RESOURCES, unless: :super_admin?}
+  validate :user_has_only_one_role, if: -> { user.present? }
+
+  class << self
+    def organizations(action)
+      resources_for(Organization, action)
+    end
+
+    def facility_groups(action)
+      resources_for(FacilityGroup, action)
+        .or(FacilityGroup.where(organization: organizations(action)))
+    end
+
+    def facilities(action)
+      resources_for(Facility, action)
+        .or(Facility.where(facility_group: facility_groups(action)))
+    end
+
+    def can?(action, model, record = nil)
+      if record&.is_a? ActiveRecord::Relation
```

Comment:
This explicit constraint is currently dubious, I'm not sure if we'll keep it or not.